extensions/typescript-language-features/src/extension.browser.ts

```diff
@@ -21,20 +21,80 @@ import { PluginManager } from './utils/plugins';
 
 class StaticVersionProvider implements ITypeScriptVersionProvider {
 
+	private configuration?: TypeScriptServiceConfiguration;
+
 	constructor(
-		private readonly _version: TypeScriptVersion
+		private readonly _browserVersion: TypeScriptVersion
 	) { }
 
-	updateConfiguration(_configuration: TypeScriptServiceConfiguration): void {
-		// noop
+	updateConfiguration(configuration: TypeScriptServiceConfiguration): void {
+		console.log(`Config updated: ${JSON.stringify(configuration, null, 4)}`);
+		this.configuration = configuration;
+	}
+
+	public get defaultVersion(): TypeScriptVersion {
+		return this.globalVersion || this.bundledVersion;
+	}
+
+	get bundledVersion() { return this._browserVersion; }
+
+	public get globalVersion(): TypeScriptVersion | undefined {
+		if (this.configuration?.globalTsdk) {
+			const globals = this.loadVersionsFromSetting(TypeScriptVersionSource.UserSetting, this.configuration.globalTsdk);
+			if (globals && globals.length) {
+				return globals[0];
+			}
+		}
+		return undefined;
+	}
+
+	public get localVersion(): TypeScriptVersion | undefined {
+		const tsdkVersions = this.localTsdkVersions;
+		if (tsdkVersions && tsdkVersions.length) {
+			return tsdkVersions[0];
+		}
+
+		const nodeVersions = this.localNodeModulesVersions;
+		if (nodeVersions && nodeVersions.length === 1) {
+			return nodeVersions[0];
+		}
+		return undefined;
+	}
+
+	public get localVersions(): TypeScriptVersion[] {
+		const allVersions = this.localTsdkVersions.concat(this.localNodeModulesVersions);
+		const paths = new Set<string>();
+		return allVersions.filter(x => {
+			if (paths.has(x.path)) {
+				return false;
+			}
+			paths.add(x.path);
+			return true;
+		});
 	}
 
-	get defaultVersion() { return this._version; }
-	get bundledVersion() { return this._version; }
+	private get localTsdkVersions(): TypeScriptVersion[] {
+		const localTsdk = this.configuration?.localTsdk;
+		return localTsdk ? this.loadVersionsFromSetting(TypeScriptVersionSource.WorkspaceSetting, localTsdk) : [];
+	}
 
-	readonly globalVersion = undefined;
-	readonly localVersion = undefined;
-	readonly localVersions = [];
+	private loadVersionsFromSetting(source: TypeScriptVersionSource, tsdkPathSetting: string): TypeScriptVersion[] {
+		if (tsdkPathSetting.startsWith("https://")) {
+			const serverPath = vscode.Uri.joinPath(vscode.Uri.parse(tsdkPathSetting), 'tsserver.js');
+			return [
+				new TypeScriptVersion(source,
+					serverPath.toString(),
+					/*DiskTypeScriptVersionProvider.getApiVersion(serverPath)*/ API.fromVersionString('4.4.1'), // TODO: Pull version form URL if possible
```

Comment:
Right now this API version is hardcoded - we could try and pull it from the URL (which could actually be anything, even a blob) or prefetch the document and search it for `ts.versionMajorMinor =` and pull out the version that way. Unsure what we want to do.

Reply:
I have [a version](https://github.com/microsoft/vscode/commit/c3fbde336e22b5ee126d7337362f2315b7d33be0) that doesn't break any of the internal APIs by using a sync `XMLHttpRequest` to inspect the server's text contents for the version string. Could be OK - definitely would want to know if that or async'ing all these getters would be preferable.

Reply:
How much work would it be to use an async call? I worry that the sync request will stall all other extension operations

Reply:
Alternative: could it say "Downloading..." and then show the ts sdk version from the API when it's finished loading and easy to access?

Reply:
> How much work would it be to use an async call? I worry that the sync request will stall all other extension operations

All the getters for different versions need to become getters for promises, and all of the functions using those getters need to become async. So pretty much everything version selector related. So long as that restructuring sounds fine, it's doable - just a large structural change for an otherwise pretty simple change. I'll finish up my local version of that for comparison - I originally bailed on it and prepped the sync request one (which shouldn't be _that_ much worse than the sync FS calls in the non browser version) once I got to needing to refactor the getters types and thus existing APIs.

Reply:
So, I have an [implementation with asyncness](https://github.com/microsoft/vscode/commit/5bdf656ad6d34838faf689df7ae4840d9892dbe7) and lemme tell you: async is toxic.